### ISSUE TITLE
SizeUnit needs to be extended for MibiBytes, KibiBytes, etc

### ISSUE THREAD

**Issue by jcalcote** [dropwizard/dropwizard#2152]
Hi guys,

I noticed today that you added support for KiB, MiB, etc a couple of years ago (https://github.com/dropwizard/dropwizard/commit/9044d44857fc1c90424303732502536c84963968), but you didn't get the semantic meaning of your extensions right. For instance, see: https://en.wikipedia.org/wiki/Gigabyte#Definition.

SizeUnit.KILOBYTE should use 1000 as the multiplier, while (the currently missing) SizeUnit.KIBIBYTE should use 1024 as the multipler. Once these additional SizeUnit values are defined, they need to be used on the proper suffixes in Size.java.

See this (10 year) old post by Jeff Atwood for a more complete picture of the differences: https://blog.codinghorror.com/gigabyte-decimal-vs-binary/. Please don't listen to the negative attitude of this article - it's 10 years old. The storage industry really does use these suffixes correctly these days and enterprise customers know the difference and expect them to be accurate.

**Comment by lutzhorn** [comment]
It should be easy to extend [`SizeUnitTest`](https://github.com/dropwizard/dropwizard/blob/master/dropwizard-util/src/test/java/io/dropwizard/util/SizeUnitTest.java), [`SizeTest`](https://github.com/dropwizard/dropwizard/blob/master/dropwizard-util/src/test/java/io/dropwizard/util/SizeTest.java), `SizeUnit`, and `Size` with this.

Why don't you open a PR?

**Comment by jcalcote** [comment]
Good idea and I will. I have one concern. The change will break backward
compatibility because currently kilobytes represents a multiple of 1024 and
according to the OSI definition KB should be a multiple of 1000 and KiB is
1024. Thoughts?

John

**Comment by lutzhorn** [comment]
Are there tests that will fail after this change? If not, tests for features that will be broken by such a change should be added first.

**Comment by isaki-x** [comment]
I doubt this is a breaking change from a drop wizard code perspective (things should still build and unit tests should still work). It may, however, break code referencing this library who are using these values to allocate buffers. If a program needs a 1 KiB buffer but is using the current Size.KILOBYTES this change will break that application as the buffer will be 24 bytes smaller.

I think this change is necessary. It may be worth pushing to 2.0 and posting a notice in the upgrade guide for migrating from 1.X to 2.0. I personally would love to see this change in 1.X series, however, I think the safer approach is called for in this instance.

**Comment by jcalcote** [comment]
The PR is open at https://github.com/dropwizard/dropwizard/pull/2154.

**Comment by joschi** [comment]
@dropwizard/committers Any strong opinions on this feature request (and the corresponding PR in #2154)?

I'm in principle for using the correct naming for units but I'm also afraid of unintended breakage in the applications of Dropwizard users.

*If* we decide to change the semantics of `SizeUnit`, we should do it now or we'll have to wait for Dropwizard 3.0.0.

**Comment by nickbabcock** [comment]
> I'm in principle for using the correct naming for units 

Agreed.

> I'm also afraid of unintended breakage in the applications of Dropwizard users

The risk, I feel here, is pretty minimal. I feel like someone who cares about the difference between kilobyte and KibiBytes (and the like) would have recognized the misnomer and coded their expectations directly using `SizeUnit.bytes` with (1000) or (1024) as the base.

**Comment by joschi** [comment]
@nickbabcock It's more about the case which @isaki outlined in his comment:

> If a program needs a 1 KiB buffer but is using the current Size.KILOBYTES this change will break that application as the buffer will be 24 bytes smaller.

**Comment by jplock** [comment]
I also agree having proper terminology is important, but I think we should plan out this change a bit more to prevent breakage (maybe introduce new terms so we can properly deprecate the old ones).  I don't think this needs to go into 2.0.0 at this time.

**Comment by nickbabcock** [comment]
I don't know if we should bend over backwards to support the use case of

```java
public void abc123() {
    byte[] myBuffer = new byte[(int)SizeUnit.KILOBYTES.toBytes(1)];
    readSomething(myBuffer);
}

private void readSomething(byte[] myBuffer) {
    if (myBuffer.length != 1024) {
        throw new IllegalArgumentException();
    }
    // ...
}
```

Seems like we would need to expend a non-negligible amount of effort to try and seamlessly transition and deprecate everyone to new terminology.

I'm also fine not rushing to a decision

**Comment by jplock** [comment]
@nickbabcock I agree.  All I'm saying is I don't think #2154 is ready to be merged in right before we release 2.0.0 (hopefully soon™)

**Comment by joschi** [comment]
@nickbabcock @jplock I've given it a try in #2686. Maybe we could merge that one for Dropwizard 2.0.0.
